include/swift/AST/ASTContext.h

```diff
@@ -414,6 +415,19 @@ class ASTContext final {
       llvm::SmallPtrSet<DerivativeAttr *, 1>>
       DerivativeAttrs;
 
+  /// For each ABI-only declaration (created with the `@abi` attribute), points
+  /// to its API-only counterpart. That is, this table maps each
+  /// `ABIAttr::abiDecl` to the declaration the `ABIAttr` is attached to.
+  ///
+  /// \seeAlso \c recordABIAttr()
+  llvm::DenseMap<Decl *, Decl *> ABIDeclCounterparts;
```

Comment:
Have you considered turning this into a cached request instead?

Reply:
This would be difficult to compute lazily because there's no way (other than this table) to walk up from an ABI-only declaration to its counterpart; you have to start from the counterpart and then walk down to the ABI-only decl. The most reliable way I've found to do that is to eagerly populate a table while parsing or deserializing. I suppose I could still create a dummy request whose `evaluate()` function returns the default output and then make `recordABIAttr()` manually cache non-default outputs, but I'm not sure if that would really buy us anything over what I'm doing here.

Reply:
I am suggesting it because we have a similar pattern with other requests that are cached manually after parse and deserialization, the primary advantage is use a common mechanism and storage.

Reply:
Changed in 413c673da4128f8a10df200899647a609c91fec1.